session/session_test.go

```diff
@@ -3259,6 +3260,138 @@ func (s *testSessionSuite2) TestSetTxnScope(c *C) {
+					{
+						GroupID: groupID,
+						Role:    placement.Leader,
+						Count:   1,
+						LabelConstraints: []placement.LabelConstraint{
+							{
+								Key:    placement.DCLabelKey,
+								Op:     placement.In,
+								Values: []string{"dc-2"},
+							},
+						},
+					},
+				},
+			}
+
+		}
+	}
+
+	// set txn_scope to global
+	tk.MustExec(fmt.Sprintf("set @@session.txn_scope = '%s';", oracle.GlobalTxnScope))
+	result := tk.MustQuery("select @@txn_scope;")
+	result.Check(testkit.Rows(oracle.GlobalTxnScope))
+	// test global txn
+	tk.MustExec("insert into t1 (c) values (1)") // in dc-1 with global scope
+	result = tk.MustQuery("select * from t1")
+	c.Assert(len(result.Rows()), Equals, 1)
+	tk.MustExec("begin")
+	txn, err := tk.Se.Txn(true)
+	c.Assert(err, IsNil)
+	c.Assert(txn.Scope(), Equals, oracle.GlobalTxnScope)
+	c.Assert(txn.Valid(), IsTrue)
+	tk.MustExec("insert into t1 (c) values (1)") // in dc-1 with global scope
+	result = tk.MustQuery("select * from t1")
+	c.Assert(len(result.Rows()), Equals, 2)
+	c.Assert(txn.Valid(), IsTrue)
+	tk.MustExec("commit")
+	result = tk.MustQuery("select * from t1")
+	c.Assert(len(result.Rows()), Equals, 2)
+	tk.MustExec("insert into t1 (c) values (101)") // in dc-2 with global scope
+	result = tk.MustQuery("select * from t1")
+	c.Assert(len(result.Rows()), Equals, 3)
+
+	// set txn_scope to local
+	tk.MustExec("set @@session.txn_scope = 'dc-1';")
+	result = tk.MustQuery("select @@txn_scope;")
+	result.Check(testkit.Rows("dc-1"))
+	// test local txn
+	tk.MustExec("insert into t1 (c) values (1)") // in dc-1 with dc-1 scope
+	result = tk.MustQuery("select * from t1")
+	c.Assert(len(result.Rows()), Equals, 4)
+	tk.MustExec("begin")
+	txn, err = tk.Se.Txn(true)
+	c.Assert(err, IsNil)
+	c.Assert(txn.Scope(), Equals, "dc-1")
+	c.Assert(txn.Valid(), IsTrue)
+	tk.MustExec("insert into t1 (c) values (1)") // in dc-1 with dc-1 scope
+	result = tk.MustQuery("select * from t1")
+	c.Assert(len(result.Rows()), Equals, 5)
+	c.Assert(txn.Valid(), IsTrue)
+	tk.MustExec("commit")
+	result = tk.MustQuery("select * from t1")
+	c.Assert(len(result.Rows()), Equals, 5)
+
+	// test wrong scope local txn
+	_, err = tk.Exec("insert into t1 (c) values (101)") // in dc-2 with dc-1 scope
+	c.Assert(err.Error(), Matches, ".*out of txn_scope.*")
+	result = tk.MustQuery("select * from t1")
+	c.Assert(len(result.Rows()), Equals, 5)
+	tk.MustExec("begin")
+	txn, err = tk.Se.Txn(true)
+	c.Assert(err, IsNil)
+	c.Assert(txn.Scope(), Equals, "dc-1")
+	c.Assert(txn.Valid(), IsTrue)
+	tk.MustExec("insert into t1 (c) values (101)") // in dc-2 with dc-1 scope
+	result = tk.MustQuery("select * from t1")
+	c.Assert(len(result.Rows()), Equals, 6)
+	c.Assert(txn.Valid(), IsTrue)
+	_, err = tk.Exec("commit")
+	result = tk.MustQuery("select * from t1")
+	c.Assert(len(result.Rows()), Equals, 5)
+	c.Assert(err.Error(), Matches, ".*out of txn_scope.*")
+	result = tk.MustQuery("select * from t1")
+	c.Assert(len(result.Rows()), Equals, 5)
```

Comment:
Can we use table-driven style to refactor these testcases? I think their structure are most the same.

Reply:
Though the structures are almost the same, the test cases are multidimensional. I tried before, but it will bring additional complexity.